### PR TITLE
AttributedString for Ti.UI.Label

### PR DESCRIPTION
**Pull request by infosia**
Fix for [TIMOB-19947](https://jira.appcelerator.org/browse/TIMOB-19947) and [TIMOB-19964](https://jira.appcelerator.org/browse/TIMOB-19964)

- Implement `Ti.UI.AttributedString`
- Implement attributedString for `Ti.UI.Label`

attributedString for `Ti.UI.Label` is a simple implementation which supports few attributes, due to mismatch with Windows document styling API.

- Only supports `FOREGROUND_COLOR` and `FONT` attribute type
- attributes can't be nested

Since Windows Xaml has its unique styling API, we might want to provide Windows-specific styling API leveraging [Windows Xaml Document API](https://msdn.microsoft.com/en-us/library/windows/apps/windows.ui.xaml.documents.run.aspx). ([TIMOB-20129](https://jira.appcelerator.org/browse/TIMOB-20129))

Test Case 1

```javascript
var win = Ti.UI.createWindow({
    backgroundColor: 'black',
});
var text = "Have you tried hyperloop yet?";
var attr = Ti.UI.createAttributedString({
    text: text,
    attributes: [
        {
            type: Ti.UI.ATTRIBUTE_FOREGROUND_COLOR,
            value: "yellow",
            range: [text.indexOf('hyperloop'), ('hyperloop').length]
        },
        {
            type: Ti.UI.ATTRIBUTE_FOREGROUND_COLOR,
            value: "red",
            range: [text.indexOf('yet'), ('yet').length]
        }
    ]
});
var label = Ti.UI.createLabel({
    attributedString: attr
});
win.add(label);
win.open();
```

Test Case 2
```javascript
var win = Titanium.UI.createWindow({
    backgroundColor: '#ddd',
});

var text =  'Bacon ipsum dolor Appcelerator Titanium rocks! sit amet fatback leberkas salami sausage tongue strip steak.';

var attr = Titanium.UI.createAttributedString({
    text: text,
    attributes: [
        {
            type: Titanium.UI.ATTRIBUTE_FONT,
            value: { fontSize: 26 },
            range: [text.indexOf('Appcelerator'), ('Appcelerator').length]
        },
        {
            type: Titanium.UI.ATTRIBUTE_FOREGROUND_COLOR,
            value: "red",
            range: [text.indexOf('Appcelerator'), ('Appcelerator').length]
        },
        {
            type: Titanium.UI.ATTRIBUTE_FOREGROUND_COLOR,
            value: "blue",
            range: [text.indexOf('Titanium'), ('Titanium').length]
        },
        {
            type: Titanium.UI.ATTRIBUTE_FOREGROUND_COLOR,
            value: "yellow",
            range: [text.indexOf('rocks!'), ('rocks!').length]
        }
    ]
});

var label = Titanium.UI.createLabel({
    left: 20,
    right: 20,
    height: Titanium.UI.SIZE,
    attributedString: attr
});

win.add(label);
win.open();
```

### CODE REVIEW

File: Source/UI/src/Label.cpp [hunk 1]
```diff
@@ -142,5 +145,91 @@ namespace TitaniumWindows
...
+		void Label::set_attributedString(const std::shared_ptr<Titanium::UI::AttributedString>& attributedString) TITANIUM_NOEXCEPT
+		{
+
+			Titanium::UI::Label::set_attributedString(attributedString);
+
+			//
+			// simple implementation for attributed string due to mismatch with Windows Style API
+			//   - only supports FOREGROUND_COLOR and FONT
+			//   - attributes can't be nested
+			// TODO: we might want to provide Windows-specific styling API leveraging Xaml Document API
```
Comment (sgtcoolguy) [inline Source/UI/src/Label.cpp]: It looks ot me like we should be able to support more than just font and fg color. https://msdn.microsoft.com/en-us/library/windows/apps/windows.ui.xaml.documents.inline.aspx

Basically you group inlines with Span and use Run to apply formatting to section of text. There are subclasses of Span that do Bold, Italic, Hyperlink and Underline.

So I think you'd basically do a Span for everything and a Run for each 'section" of formatted text?

Anyways, we could use the Hyperlink and Underline versions to do Titanium.UI.ATTRIBUTE_UNDERLINES_STYLE
Titanium.UI.ATTRIBUTE_LINK ?

Titanium.UI.ATTRIBUTE_KERN is equivalent to CharacterSpacing on the Span/Run:  https://msdn.microsoft.com/en-us/library/windows/apps/windows.ui.xaml.documents.textelement.characterspacing.aspx

Comment (sgtcoolguy) [inline Source/UI/src/Label.cpp]: ATTRIBUTE_EXPANSION roughly correlates to https://msdn.microsoft.com/en-us/library/windows/apps/windows.ui.text.fontstretch.aspx

Comment (sgtcoolguy) [inline Source/UI/src/Label.cpp]: ATTRIBUTE_WRITING_DIRECTION roughly correlates to https://msdn.microsoft.com/en-us/library/windows/apps/windows.ui.xaml.documents.run.flowdirection.aspx

Comment (sgtcoolguy) [inline Source/UI/src/Label.cpp]: Ligature = https://msdn.microsoft.com/en-us/library/windows/apps/windows.ui.xaml.documents.typography.standardligatures.aspx ?

Comment (sgtcoolguy) [inline Source/UI/src/Label.cpp]: Another kerning reference: https://msdn.microsoft.com/en-us/library/windows/apps/windows.ui.xaml.documents.typography.setkerning.aspx

**Comment by infosia** [comment]
Merging, will push more properties in subsequent PR.

**Pull request merged**